docs/adr/0007-preventing-version-mismatches.md

```diff
@@ -0,0 +1,233 @@
+    coresponding release for all the other crates.
+  - Even releases with breaking changes don't always require corresponding releases in the other crates.  For example,
+    if we could update how `uniffi-bindgen-kotlin` generates the class hierarchy for Kotlin errors, without needing to
+    change `uniffi` or `uniffi-bindgen`.
+
+This means that our current method of using exact string matching for versions is no longer going to work.
+
+Finally, there can only be one `uniffi-bindgen` and `uniffi-bindgen-[language]`, installed system wide.  This makes it
+difficult for a developer to work on crates that depend on different UniFFI versions.
+
+We've built a decent solution for the scaffolding code: [the builtin-bingen
+feature](https://mozilla.github.io/uniffi-rs/tutorial/Rust_scaffolding.html#avoiding-version-mismatches-between-uniffi-core-and-uniffi-bindgen).
+However, there's no similar solution for the bindings code.
+
+application-services has a custom solution for the bindings generation.  It contains the
+[embedded-uniffi-bindgen](https://github.com/mozilla/application-services/tree/main/tools/embedded-uniffi-bindgen) crate
+which contains very
+[simple binary](https://github.com/mozilla/application-services/blob/main/tools/embedded-uniffi-bindgen/src/main.rs)
+that just runs `uniffi-bindgen`. This allows us to control the `uniffi-bindgen` version using the `Cargo.toml` for this
+crate, rather than relying on the system installed version.
+
+### Breaking changes to the UDL syntax
+
+A related issue is breaking changes to the UDL syntax, for example the change the renamed `[Wrapped]` to `[Custom]`.
+These kind of changes can lead to challenges with scaffolding and bindings generation.  For example, users could get
+into a situtation where `uniffi-bindgen` can parse a UDL file, but `uniffi-bindgen-kotlin` can't.
+
+However, this issue seems less important since a) the errors are easier to diagnose and fix and b) we have historically
+only introduced breaking changes like this for "experimental" features like custom/wrapped types.
+
+This ADR will ignore this issue and only focus on compatibility between the generated scaffolding and bindings code.
+
+## Decision Drivers
+
+- UniFFI should have a friendly developer experience.
+  - If possible, developers shouldn't have to worry or know about this issue.
+  - If that's not possible, then it should be easy for them to identify and fix the issue.
+- We need to support external bindings generators and allow those projects to
+  release new versions without coordinating with the UniFFI team.
+
+## Considered Options
+
+### **[Requirement] More flexible versioning**
+
+As mentioned in the context, once we have external bindings generators we can't rely on a simply compare it's version
+string to `uniffi`.  A more flexible versioning policy is a requirement for all other options discussed here.
+
+I think we can use semver for this:
```

Comment:
Is this referring specifically to the version number of the `uniffi` crate, the bindings-generator crates, etc? Or, is it a separate version number specifically for the details of the FFI interface, shared between all the different crates in the ecosystem?

If the intent is to use the package version number for this, there may be a nontrivial risk of "false positive" version bumps, where we e.g. rev the major version number of `uniffi` because of some consumer-facing API change in the tooling or integration, causing incompatibility with bindings generators despite no actual change in the FFI layer.

Reply:
Hey Ryan! I think the intent here was to use the crate versions, with the implication being that uniffi and the bindings would need to stay in a kind of lock-step regarding the major version number. You are correct that "false positives" would be likely in this scenario - I wonder if we can come up with examples of this, because I could imagine that many tooling or integration changes would indeed impact the bindings generators in some way even if they didn't change the ffi layer. Obviously there are cases where this wouldn't be true too.

A separate version for the ffi layer would be good, but then we couldn't rely on cargo - OTOH though, maybe that's not a big problem in practice - the fact that a human might need to determine the exact version of the "foo" bindings which interoperate with version x.y.z of uniffi might be ok.